File: app/src/main/java/com/mgaetan89/showsrage/helper/ImageLoader.java
package com.mgaetan89.showsrage.helper;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;
import android.support.v7.graphics.Palette;
import android.widget.ImageView;

import com.bumptech.glide.BitmapRequestBuilder;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.BitmapImageViewTarget;

public abstract class ImageLoader {
	public interface OnImageResult {
		void onImageError(ImageView imageView, @Nullable Exception exception, @Nullable Drawable errorDrawable);

		void onImageReady(ImageView imageView, Bitmap resource);
	}

	public static void load(@Nullable ImageView imageView, String url, boolean circleTransform) {
		load(imageView, url, circleTransform, null, null);
	}

	public static void load(@Nullable ImageView imageView, String url, boolean circleTransform, @Nullable final Palette.PaletteAsyncListener paletteListener, @Nullable final OnImageResult onImageResult) {
		if (imageView == null) {
			return;
		}

		BitmapRequestBuilder<String, Bitmap> glide = Glide.with(imageView.getContext())
				.load(url)
				.asBitmap()
				.diskCacheStrategy(DiskCacheStrategy.ALL);

		if (circleTransform) {
			BitmapPool bitmapPool = Glide.get(imageView.getContext()).getBitmapPool();

			glide.transform(new GlideCircleTransformation(bitmapPool));
		}

		glide.into(new BitmapTarget(imageView, paletteListener, onImageResult));
	}

	private static final class BitmapTarget extends BitmapImageViewTarget {
		private OnImageResult onImageResult;
		private Palette.PaletteAsyncListener paletteListener;

		private BitmapTarget(ImageView view, Palette.PaletteAsyncListener paletteListener, OnImageResult onImageResult) {
			super(view);

			this.onImageResult = onImageResult;
			this.paletteListener = paletteListener;
		}

		@Override
		public void onLoadFailed(Exception e, Drawable errorDrawable) {
			super.onLoadFailed(e, errorDrawable);

			if (this.onImageResult != null) {
				this.onImageResult.onImageError(this.getView(), e, errorDrawable);
			}
		}

		@Override
		public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
			super.onResourceReady(resource, glideAnimation);

			if (this.onImageResult != null) {
				this.onImageResult.onImageReady(this.getView(), resource);
			}

			if (this.paletteListener != null) {
				new Palette.Builder(resource).generate(this.paletteListener);
			}
		}
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/helper/ImageLoader.kt
package com.mgaetan89.showsrage.helper

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.support.annotation.IdRes
import android.support.annotation.WorkerThread
import android.support.v7.graphics.Palette
import android.widget.ImageView
import android.widget.RemoteViews
import com.bumptech.glide.Glide
import com.bumptech.glide.RequestBuilder
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.BitmapImageViewTarget
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.transition.Transition

object ImageLoader {
	interface OnImageResult {
		fun onImageError(imageView: ImageView, errorDrawable: Drawable?)

		fun onImageReady(imageView: ImageView, resource: Bitmap)
	}

	fun getBitmap(context: Context, url: String?, circleTransform: Boolean)
			= this.getGlideInstance(context, url, circleTransform).submit(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)

	fun load(imageView: ImageView?, url: String?, circleTransform: Boolean) {
		this.load(imageView, url, circleTransform, null, null)
	}

	fun load(imageView: ImageView?, url: String?, circleTransform: Boolean, paletteListener: Palette.PaletteAsyncListener?, onImageResult: OnImageResult?) {
		val context = imageView?.context ?: return

		this.getGlideInstance(context, url, circleTransform).into(BitmapTarget(imageView, paletteListener, onImageResult))
	}

	@WorkerThread
	fun load(context: Context, remoteViews: RemoteViews, @IdRes viewId: Int, url: String?, circleTransform: Boolean) {
		this.getBitmap(context, url, circleTransform).let {
			remoteViews.setImageViewBitmap(viewId, it.get())

			Glide.with(context.applicationContext).clear(it)
		}
	}

	private fun getGlideInstance(context: Context, url: String?, circleTransform: Boolean): RequestBuilder<Bitmap> {
		val requestOptions = RequestOptions()
			.diskCacheStrategy(DiskCacheStrategy.ALL)
			.apply {
				if (circleTransform) {
					this.circleCrop()
				}
			}

		return Glide.with(context.applicationContext)
			.asBitmap()
			.load(url)
			.apply(requestOptions)
	}

	private class BitmapTarget(view: ImageView, val paletteListener: Palette.PaletteAsyncListener?, val onImageResult: OnImageResult?) : BitmapImageViewTarget(view) {
		override fun onLoadFailed(errorDrawable: Drawable?) {
			super.onLoadFailed(errorDrawable)

			this.onImageResult?.onImageError(this.view, errorDrawable)
		}

		override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
            this.view.setImageBitmap(resource)

			this.onImageResult?.onImageReady(this.view, resource)

			this.paletteListener?.let {
				Palette.Builder(resource).generate(it)
			}
		}
	}
}
